Bottom: a33ac57ae0539167c42d823b36dc16f202b7e2b2
Top:    b3481d2ba452e43d2268bd7771737cff078d0233
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-07-17 17:48:30 +0800

Refresh of vhost_net-split-out-datacopy

---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index 10e4f7a..af938f4 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -495,10 +495,12 @@ static int get_tx_bufs(struct vhost_net *net,
 
 	ret = vhost_net_tx_get_vq_desc(net, vq, out, in, busyloop_intr);
 
+	if (ret < 0 || ret == vq->num)
+		return ret;
+
 	if (*in) {
 		vq_err(vq, "Unexpected descriptor format for TX: out %d, int %d\n", *out, *in);
-		ret = -EFAULT;
-		goto err;
+		return -EFAULT;
 	}
 
 	/* Sanity check */
@@ -507,14 +509,10 @@ static int get_tx_bufs(struct vhost_net *net,
 		vq_err(vq, "Unexpected header len for TX: "
 			"%zd expected %zd\n",
 			*len, nvq->vhost_hlen);
-		ret = -EFAULT;
-		goto err;
+		return -EFAULT;
 	}
 
 	return ret;
-err:
-	*len = 0;
-	return ret;
 }
 
 static bool tx_can_batch(struct vhost_virtqueue *vq, size_t total_len)
